ampd/src/grpc/reqs.rs

```diff
@@ -1,41 +1,60 @@
 use std::str::FromStr;
 
-use ampd_proto::{BroadcastRequest, ContractStateRequest, SubscribeRequest};
+use ampd_proto::{
+    Algorithm, BroadcastRequest, ContractStateRequest, KeyId, KeyRequest, SignRequest,
+    SubscribeRequest,
+};
 use axelar_wasm_std::nonempty;
 use cosmrs::Any;
-use error_stack::{ensure, report, Report, Result, ResultExt};
+use error_stack::{bail, ensure, report, Report, Result, ResultExt};
 use report::ResultCompatExt;
 use thiserror::Error;
 use tonic::Request;
 
 use crate::types::TMAddress;
-use crate::PREFIX;
+use crate::{tofnd, PREFIX};
 
-pub fn validate_subscribe(req: Request<SubscribeRequest>) -> Result<EventFilters, Error> {
-    let SubscribeRequest {
-        filters,
-        include_block_begin_end,
-    } = req.into_inner();
+pub trait Validate {
+    type Output;
 
-    (filters, include_block_begin_end).try_into()
+    fn validate(self) -> Result<Self::Output, Error>;
 }
 
-pub fn validate_broadcast(req: Request<BroadcastRequest>) -> Result<Any, Error> {
-    req.into_inner()
-        .msg
-        .ok_or(report!(Error::EmptyBroadcastMsg))
+impl Validate for Request<SubscribeRequest> {
+    type Output = EventFilters;
+
+    fn validate(self) -> Result<Self::Output, Error> {
+        let SubscribeRequest {
+            filters,
+            include_block_begin_end,
+        } = self.into_inner();
+
+        (filters, include_block_begin_end).try_into()
+    }
+}
+
+impl Validate for Request<BroadcastRequest> {
+    type Output = Any;
+
+    fn validate(self) -> Result<Self::Output, Error> {
+        self.into_inner()
+            .msg
+            .ok_or(report!(Error::EmptyBroadcastMsg))
+    }
 }
 
-pub fn validate_contract_state(
-    req: Request<ContractStateRequest>,
-) -> Result<(TMAddress, Vec<u8>), Error> {
-    let ContractStateRequest { contract, query } = req.into_inner();
+impl Validate for Request<ContractStateRequest> {
+    type Output = (TMAddress, Vec<u8>);
 
-    ensure!(!query.is_empty(), Error::InvalidQuery);
-    let _: serde_json::Value =
-        serde_json::from_slice(&query).change_context(Error::InvalidQuery)?;
+    fn validate(self) -> Result<Self::Output, Error> {
+        let ContractStateRequest { contract, query } = self.into_inner();
 
-    Ok((validate_address(&contract)?, query))
+        ensure!(!query.is_empty(), Error::InvalidQuery);
+        let _: serde_json::Value =
+            serde_json::from_slice(&query).change_context(Error::InvalidQuery)?;
```

Comment:
this needs to become generic, right?

Reply:
yeah, validate that the query bytes can be deserialized to valid json